### PR TITLE
[tensorflow] | [test] | [benchmark] | [sagemaker] feature: Add Tensorflow sagemaker benchmark tests

### PR DESCRIPTION
**Pull request by saimidu**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

*Description:*
Added tests for SageMaker Benchmarking of TensorFlow Training images

*Tests run:*
Tested by triggering SageMaker tests locally through `test/testrunner.py`

*DLC image/dockerfile:*

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow1/singletrain_cpu.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow1/singletrain_cpu.sh:1]: This script is unchanged from the original.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow1/singletrain_gpu.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow1/singletrain_gpu.sh:1]: This script is unchanged from the original.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow2/singletrain_cpu.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow2/singletrain_cpu.sh:1]: This script is unchanged from the original.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow2/singletrain_gpu.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tensorflow2/singletrain_gpu.sh:1]: This script is unchanged from the original.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tf_sm_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import argparse
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/resources/tf_sm_benchmark.py:1]: This script is a combination of https://github.com/akhilmehra/sagemakerbenchmarking/blob/tf2/src/startit_tf_gpu.py and https://github.com/akhilmehra/sagemakerbenchmarking/blob/tf2/src/startit_tf_cpu.py

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import os
```
Comment (saimidu) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py:1]: This script is almost completely new.

File: test/test_utils/__init__.py [hunk 7]
```diff
@@ -321,3 +322,35 @@ def get_dlc_images():
...
+
+
+def setup_sm_benchmark_tf_train_env(resources_location, setup_tf1_env, setup_tf2_env):
+    """
+    Create a virtual environment for benchmark tests if it doesn't already exist, and download all necessary scripts
+    :param resources_location: <str> directory in which test resources should be placed
+    :param setup_tf1_env: <bool> True if tf1 resources need to be setup
+    :param setup_tf2_env: <bool> True if tf2 resources need to be setup
+    :return: absolute path to the location of the virtual environment
+    """
```
Comment (saimidu) [inline test/test_utils/__init__.py:334]: This function needs to execute in order to perform a one-time setup before TF SM benchmark tests can be run.

File: test/testrunner.py [hunk 8]
```diff
@@ -179,9 +189,17 @@ def main():
...
     elif specific_test_type == "sagemaker":
-        run_sagemaker_tests(
-            [image for image in standard_images_list if not ("tensorflow-inference" in image and "py2" in image)]
-        )
+        if benchmark_mode:
+            report = os.path.join(os.getcwd(), "test", f"{test_type}.xml")
+            os.chdir(os.path.join("test", "dlc_tests"))
+
+            setup_sm_benchmark_env(dlc_images, test_path)
+            pytest_cmd = ["-s", "-rA", test_path, f"--junitxml={report}", "-n=auto", "-o", "norecursedirs=resources"]
```
Comment (saimidu) [inline test/testrunner.py:197]: "-o norecursedirs" ensures that pytest does not recursive execute the tests in any clones of other GitHub repositories that have been done for the sake of the tests.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+from test.test_utils import BENCHMARK_RESULTS_S3_BUCKET
+
+
+@pytest.mark.parametrize("num_nodes", [1, 4], indirect=True)
+def test_tensorflow_sagemaker_training_performance(tensorflow_training, num_nodes, region):
+
+    # This sleep has been inserted because all the parametrized training jobs are automatically created
+    # by SageMaker with the same name, due to being started around the same time, and with the same image uri.
+    time.sleep(random.Random(x=f"{tensorflow_training}{num_nodes}").random() * 60)
```
Comment (TusharKanekiDey) [inline test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py:18]: Can you please explain the functionality of this line?

Comment (saimidu) [reply test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py:18]: We can have TensorFlow Training images for CPU/GPU, Py2/Py3, TF-1 and TF-2, all of which can be performance tested using the same test function.

It is unlikely that both TF-1 and TF-2 images will be tested in the same process, but it is highly likely that 2-4 images (CPU/GPU-Py2/Py3/Py37) will be tested as a part of the same CodeBuild job.

When these threads are started within moments of one-another, the SageMaker Training Job that they create for the benchmark test will also be created around the same time. Each of these jobs has a "unique" name, that is automatically generated by SageMaker based on the time at which the test was started. However, this unique name only takes the repository name into account, i.e., "tensorflow-training", and not the actual image tag, i.e., "2.2.0-cpu/gpu-py37/...".

Therefore, I saw 3-4 attempts to run tests on all images simultaneously fail, because the name of the Training Job created by SageMaker was the same.

This random wait time between 0 and 60 seconds ensures that each image waits for a different amount of time before starting its test.

`Random(x="...")` is an instance of the `Random` class, where the unique combination of (image_uri + num_nodex) is given as a seed to ensure the time waited is always different.

File: test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                          f"--python {py_version} "
+                          f"--region {region} "
+                          f"> {log_file}",
+                          warn=True, echo=True)
+
+        if not (run_out.ok or run_out.return_code == 124):
+            target_upload_location = os.path.join(target_upload_location, "failure_log")
+
+        ctx.run(f"aws s3 cp {log_file} {os.path.join(target_upload_location, log_file)}")
+        ctx.run(f"cat {log_file}", echo=True)
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py:54]: This is true. I will make the change to provide just a link to the actual test results. It doesn't look like the last few lines of the file have a lot of useful information about the performance results. We'll need to print about 30 lines of the end of the log file to see useful information.

[the remark this answers is not in the document]
Comment (YYStreet) [reply test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py:54]: Sorry I did some modification on the comments because I forgot to save it.
We could also consider to do post-process from origin log file to extract the most interesting part to print to codebuild. Of course we might not need to do it.
